### PR TITLE
Processed feedback for Software/Infrastructure Requirements

### PR DESCRIPTION
**Pull request by proycon**
This PR suggest various improvements to the Software Requirements and Infrastructure Requirements, based on the feedback during the CLARIAH Tech Day last wednesday. Please review carefully, see individual commits messages for incremental changes. Also see #102 for further context.

Issue #109 is still open and **NOT** addressed yet in this PR (requires some further investigation).

### CODE REVIEW

**Review by jblom — APPROVED** [review]
@proycon the text is much less ambiguous now. All changes are improvements

**Comment by proycon** [comment]
@ddedoer: Kan jij je er ook in vinden zo?

**Comment by proycon** [comment]
@ddeboer: Kan jij je er ook in vinden zo?

File: requirements/infrastructure-requirements.md [hunk 1]
```diff
@@ -161,19 +161,22 @@ Logs can be made available through a command-line and/or a web interface.
 
 ### 7. The infrastructure *MUST* _automatically build_ the application. ([AUTO](#auto))
 
-When commits are pushed to the application source code or releases are tagged, the infrastructure must automatically
-(re)build the application, producing an OCI container build artefact.
+When software is released, through tagging in the Version Control System (VCS), the infrastructure must automatically
+(re)build the application, producing an OCI container build artefact. This facilitates *continuous deployment*.
```
Comment (ddeboer) [inline requirements/infrastructure-requirements.md:165]:
```suggestion
(re)build the application, producing an OCI container build artefact. This facilitates *continuous delivery and deployment*.
```

File: requirements/infrastructure-requirements.md [hunk 2]
```diff
@@ -267,7 +270,6 @@ In the case of horizontal scaling (scaling in and out) this requires load-balanc
 
 If the infrastructure is able to directly serve static files, without configuring a web server first,
 this speeds up delivery.
-Moreover, caching these files in a third-party Content Delivery Network (CDN) will reduce latency for users.
```
Comment (ddeboer) [inline requirements/infrastructure-requirements.md:270]: Why did you remove this? I added this originally to illustrate the idea that infrastructure, at it is defined in the introduction, is spread over a number of components (including an external CDN) rather than a single configuration on some server.

File: requirements/software-requirements.md [hunk 3]
```diff
@@ -62,7 +65,7 @@ Periodic releases of the software *SHOULD* be released with a clear version numb
    platforms such a GitHub, GitLab, Bitbucket.
 3. Each release *must* be accompanied by release notes describing on a more
    general level what is new in the release. Again, the release mechanisms by
-   major CVS platforms provide for this. Alternatively you *MAY* use a ``NEWS``
+   major VCS platforms provide for this. Alternatively you *MAY* use a ``NEWS``
```
Comment (ddeboer) [inline requirements/software-requirements.md:68]: 😅

File: requirements/software-requirements.md [hunk 4]
```diff
@@ -77,32 +80,53 @@ This is a corollary of [Infrastructure Requirement 4](infrastructure-requirement
...
 * Java libraries *SHOULD* be packaged for and submitted to Maven Central, installable through ``mvn``.
-* Rust software *SHOULD* be packaged for and submitted to [crates.io](https://crates.io/), installable through ``cargo``.
-* NodeJS software *SHOULD* be packaged for and submitted to [npm](https://www.npmjs.com/), installable through ``npm``.
+* Rust libraries and command-line tools *SHOULD* be packaged for and submitted to [crates.io](https://crates.io/), installable through ``cargo``.
+* NodeJS libraries *SHOULD* be packaged for and submitted to [npm](https://www.npmjs.com/), installable through ``npm``.
 * C/C++ software has no specific ecosystem for packaging. The use of
-  distribution-specific packages is *RECOMMENDED*, targetting specific Linux
-  distributions. The use of a standardized build system such as the autotools,
+  distribution-specific packages as described further below is then *RECOMMENDED*
+  The use of an standardized build system such as the autotools,
```
Comment (ddeboer) [inline requirements/software-requirements.md:97]:
```suggestion
  The use of a standardized build system such as the autotools,
```

File: requirements/software-requirements.md [hunk 5]
```diff
@@ -77,32 +80,53 @@ This is a corollary of [Infrastructure Requirement 4](infrastructure-requirement
...
 
 This step may combine both the building of software (compilation) as well the
-installation. This rule shall only be deviated from if there is no suitable
+installation. If any building/compilation is performed, then the build process
+*MUST* make use of an *automated* and *standardized* build system. 
+
+Packaging for a package manager shall only be deviated from if there is no suitable
 packaging ecosystem for the software (C/C++) or if the software is not a fit
 for the packaging ecosystem, such as the software being too high-level (e.g.  a
 complex web application) and consisting of too many components for any
```
Comment (ddeboer) [inline requirements/software-requirements.md:109]: Not necessarily complex, though. Why would you want to publish a web application as a package? (Web) _applications_ are typically published as OCI images, only the reusable _libraries_ that they depend on should be published as packages.

File: requirements/software-requirements.md [hunk 6]
```diff
@@ -233,6 +257,7 @@ local Mac or Windows machine.
    IR6).
 5. Application data (state) that needs to be persistent between runs *MUST* be stored separate from the container (e.g. in a
    mounted volume) (See point 3 of the [Infrastructure Requirements (IR)](infrastructure-requirements.md))
+6. The container build process *SHOULD* make use proper packaging ecosystems for installing its individual components. Providing a container *SHOULD NOT* be considered a substitute for component packaging (see point 6, which is to be considered as complementary to or a prerequisite for this one).
```
Comment (ddeboer) [inline requirements/software-requirements.md:260]:
```suggestion
6. The container build process *SHOULD* use proper packaging ecosystems for installing its individual components. Providing a container *SHOULD NOT* be considered a substitute for component packaging (see point 6, which is to be considered as complementary to or a prerequisite for this one).
```

Comment (ddeboer) [reply requirements/software-requirements.md:260]: I don’t completely agree with this. To me, the distinction applications (published as container images) and libraries (published as language-specific packages) makes more sense.

File: requirements/software-requirements.md [hunk 7]
```diff
@@ -254,6 +279,7 @@ for this communication. Instruction can be found [here](https://github.com/CLARI
 
 1. [OpenAPI](https://spec.openapis.org/oas/latest.html) (aka Swagger), WADL, and CLAM are *SUGGESTED* as possible interface description languages.
 2. The specification endpoint *SHOULD NOT* be hindered by any authentication barriers.
+3. The specification endpoint *SHOULD* be automatically generated
```
Comment (ddeboer) [inline requirements/software-requirements.md:282]:
```suggestion
3. The specification endpoint *SHOULD* be automatically generated.
```

File: requirements/infrastructure-requirements.md (hunk 2 above)
Comment (proycon) [reply requirements/infrastructure-requirements.md:270]: I think there is a conflict here between this phrase and point 23 of the Software Requirements. Recommending third-party content-delivery networks leads to cross-origin requests and users might be tracked in some form by those networks (like CloudFlare's).

(Ideally I'd rather even recommend limiting cross-origin requests in general, for the same reason)

File: requirements/software-requirements.md (hunk 5 above)
Comment (proycon) [reply requirements/software-requirements.md:109]: Right, perhaps is still worded a bit too strongly and we should removed the adjective *complex* here and make clear that web applications are exempted to prevent confusion.

(Technically, there is still also a place for web applications in Linux distributions of course, for example, Alpine Linux packages things like nextcloud, but I think for CLARIAH that would be overkill and may conflict with the container model which we require)

File: requirements/infrastructure-requirements.md (hunk 2 above)
Comment (jblom) [reply requirements/infrastructure-requirements.md:270]: tend to agree with @proycon on CDN's and cross-origin requests.

Maybe it 's good to think/talk about CLARIAH's stance towards using third-party software (which parties can we trust and which do we try to avoid).

If we are clear on this we can still illustrate the requirments with examples using third party software we support.

Also deciding NOT going into this (because it might be too political/frustrating) is good to get consensus about.

Comment (proycon) [reply requirements/infrastructure-requirements.md:270]: > Also deciding NOT going into this (because it might be too political/frustrating) is good to get consensus about.

Indeed, it might be best to be a bit neutral on this in the text. Which is what I think we get now we remove this CDN suggestion (and not have a clause discouraging cross-origins like I'd ideally like). Then people can still go either way.

Comment (ddeboer) [reply requirements/infrastructure-requirements.md:270]: Agreed!

File: requirements/software-requirements.md (hunk 5 above)
Comment (proycon) [reply requirements/software-requirements.md:109]: Should be resolved with https://github.com/CLARIAH/clariah-plus/commit/3c79189412b4b4400182337aaead0a0b696564c3 now

File: requirements/infrastructure-requirements.md (hunk 2 above)
Comment (jblom) [reply requirements/infrastructure-requirements.md:270]: +1!
